File: src/cleansia_android/customer-app/src/main/java/cz/cleansia/customer/features/booking/BookingSuccessViewModel.kt
package cz.cleansia.customer.features.booking

import android.util.Log
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cz.cleansia.customer.core.orders.OrderApi
import cz.cleansia.customer.core.orders.OrderDetailDto
import cz.cleansia.customer.core.orders.OrderRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * UI state for [BookingSuccessScreen]'s enrichment block.
 *
 * The screen always renders its celebratory header + CTAs; this state only
 * drives the optional enrichment (confirmation code, arrival window, address,
 * total). On [Error], the screen silently drops the block rather than
 * surfacing a "we couldn't load your order" message — the user just booked,
 * the nav args already carry the confirmation code, and the OrdersTab will
 * show the full record on the next visit.
 */
sealed interface BookingSuccessUiState {
    data object Loading : BookingSuccessUiState
    data class Loaded(val order: OrderDetailDto) : BookingSuccessUiState
    data object Error : BookingSuccessUiState
}

/**
 * Fetches the newly-created order's detail once for the success screen so
 * the user sees their arrival window, address and total without having to
 * bounce to the Orders tab.
 *
 * Bypasses [OrderRepository.getById] on purpose: that helper surfaces a
 * snackbar on failure (by design for the Orders-tab flow), which would
 * muddy the post-payment celebration. Here we hit [OrderApi] directly and
 * fall back to a silent [BookingSuccessUiState.Error] — the screen degrades
 * gracefully.
 *
 * Side effect: pokes [OrderRepository.refresh] so when the user navigates
 * to the Orders tab afterwards their new order is already in the list.
 * This is deliberately fire-and-forget; the tab also re-fetches when the
 * repo is empty, so a failed refresh here is harmless.
 */
@HiltViewModel
class BookingSuccessViewModel @Inject constructor(
    private val orderApi: OrderApi,
    private val orderRepository: OrderRepository,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    private val orderId: String? = savedStateHandle.get<String>("orderId")

    private val _state = MutableStateFlow<BookingSuccessUiState>(BookingSuccessUiState.Loading)
    val state: StateFlow<BookingSuccessUiState> = _state.asStateFlow()

    init {
        // Pre-warm the Orders list so the new order is visible the moment the
        // user lands on the Orders tab. Failure is silent — the tab refreshes
        // itself when it's empty.
        viewModelScope.launch {
            orderRepository.refresh()
        }
        load()
    }

    private fun load() {
        val id = orderId
        if (id.isNullOrBlank()) {
            _state.value = BookingSuccessUiState.Error
            return
        }
        _state.value = BookingSuccessUiState.Loading
        viewModelScope.launch {
            val dto = try {
                val resp = orderApi.getById(id)
                if (resp.isSuccessful) resp.body() else null
            } catch (t: Throwable) {
                Log.w(TAG, "BookingSuccess: failed to fetch order $id", t)
                null
            }
            _state.value = if (dto != null) {
                BookingSuccessUiState.Loaded(dto)
            } else {
                Log.w(TAG, "BookingSuccess: empty order body for $id — degrading to generic success")
                BookingSuccessUiState.Error
            }
        }
    }

    private companion object {
        const val TAG = "BookingSuccessVM"
    }
}

File: src/cleansia_ios/CleansiaCustomer/Sources/Features/Booking/BookingSuccessViewModel.swift
import CleansiaCustomerApi
import Foundation

/// The `BookingSuccessUiState` (Android) parity: the screen always renders the
/// celebratory header + code pill; this state only drives the enrichment
/// (summary card + timeline detail). `.error` degrades silently — the success
/// moment is never blocked on a failed fetch.
enum BookingSuccessUiState {
    case loading
    case loaded(OrderItem)
    case error
}

@MainActor
final class BookingSuccessViewModel: ObservableObject {
    @Published private(set) var state: BookingSuccessUiState = .loading

    private let orderId: String
    private let fetch: @Sendable (String) async -> OrderItem?
    private let warmOrders: @Sendable () async -> Void
    private var didLoad = false

    init(
        orderId: String,
        fetch: @escaping @Sendable (String) async -> OrderItem?,
        warmOrders: @escaping @Sendable () async -> Void = {}
    ) {
        self.orderId = orderId
        self.fetch = fetch
        self.warmOrders = warmOrders
    }

    var order: OrderItem? {
        if case let .loaded(order) = state { return order }
        return nil
    }

    /// The freshly-loaded code wins over the nav-arg one (the backend may trim
    /// whitespace); the fallback keeps the pill rendering in loading/error.
    func effectiveCode(fallback: String) -> String {
        guard let code = order?.confirmationCode, !code.isBlank else { return fallback }
        return code
    }

    func load() async {
        guard !didLoad else { return }
        didLoad = true
        async let warm: Void = warmOrders()
        if orderId.isBlank {
            state = .error
        } else if let order = await fetch(orderId) {
            state = .loaded(order)
        } else {
            state = .error
        }
        await warm
    }
}
